subsys/net/ip/net_shell.c

```diff
@@ -1887,6 +1888,400 @@ static int cmd_net_dns(const struct shell *shell, size_t argc, char *argv[])
+
+		info = net_addr_ntop(AF_INET, &data->requested_ip, extra_info,
+				     extra_info_len);
+#endif
+		break;
+	case NET_EVENT_IPV4_DHCP_STOP:
+		*desc = "DHCPv4";
+		*desc2 = "stop";
+		break;
+	}
+
+	return info;
+}
+
+static const char *get_l4_desc(uint32_t event)
+{
+	static const char *desc = "<unknown event>";
+
+	switch (event) {
+	case NET_EVENT_L4_CONNECTED:
+		desc = "connected";
+		break;
+	case NET_EVENT_L4_DISCONNECTED:
+		desc = "disconnected";
+		break;
+	case NET_EVENT_DNS_SERVER_ADD:
+		desc = "DNS server add";
+		break;
+	case NET_EVENT_DNS_SERVER_DEL:
+		desc = "DNS server del";
+		break;
+	}
+
+	return desc;
+}
+
+/* We use a separate thread in order not to do any shell printing from
+ * event handler callback (to avoid stack size issues).
+ */
+static void event_mon_handler(const struct shell *shell)
+{
+	char extra_info[NET_IPV6_ADDR_LEN];
+	struct event_msg msg;
+
+	net_mgmt_init_event_callback(&l2_cb, event_handler,
+				     MONITOR_L2_MASK);
+	net_mgmt_add_event_callback(&l2_cb);
+
+	net_mgmt_init_event_callback(&l3_ipv4_cb, event_handler,
+				     MONITOR_L3_IPV4_MASK);
+	net_mgmt_add_event_callback(&l3_ipv4_cb);
+
+	net_mgmt_init_event_callback(&l3_ipv6_cb, event_handler,
+				     MONITOR_L3_IPV6_MASK);
+	net_mgmt_add_event_callback(&l3_ipv6_cb);
+
+	net_mgmt_init_event_callback(&l4_cb, event_handler,
+				     MONITOR_L4_MASK);
+	net_mgmt_add_event_callback(&l4_cb);
+
+	while (net_event_shutting_down == false) {
+		const char *layer_str = "<unknown layer>";
+		const char *desc = "", *desc2 = "";
+		char *info = NULL;
+		uint32_t layer;
+
+		(void)k_msgq_get(&event_mon_msgq, &msg, K_FOREVER);
+
+		if (msg.iface == NULL && msg.event == 0 && msg.len == 0) {
+			/* This is the stop message */
+			continue;
+		}
+
+		layer = NET_MGMT_GET_LAYER(msg.event);
+
+		if (layer == NET_MGMT_LAYER_L2) {
+			layer_str = "L2";
+			desc = get_l2_desc(msg.event);
+		} else if (layer == NET_MGMT_LAYER_L3) {
+			layer_str = "L3";
+			info = get_l3_desc(&msg, &desc, &desc2,
+					   extra_info, NET_IPV6_ADDR_LEN);
+		} else if (layer == NET_MGMT_LAYER_L4) {
+			layer_str = "L4";
+			desc = get_l4_desc(msg.event);
+		}
+
+		PR_INFO("EVENT: %s [%d] %s%s%s%s%s\n", layer_str,
+			net_if_get_by_iface(msg.iface), desc,
+			desc2 ? " " : "", desc2 ? desc2 : "",
+			info ? " " : "", info ? info : "");
+	}
+
+	net_mgmt_del_event_callback(&l2_cb);
+	net_mgmt_del_event_callback(&l3_ipv4_cb);
+	net_mgmt_del_event_callback(&l3_ipv6_cb);
+	net_mgmt_del_event_callback(&l4_cb);
+
+	net_event_monitoring = false;
+	net_event_shutting_down = false;
+
+	PR_INFO("Network event monitoring %s.\n", "disabled");
+}
+#endif
+
+static int cmd_net_events_on(const struct shell *shell, size_t argc,
+			     char *argv[])
+{
+#if defined(CONFIG_NET_MGMT_EVENT_MONITOR)
+	k_tid_t tid;
+
+	if (net_event_monitoring) {
+		PR_INFO("Network event monitoring is already %s.\n",
+			"enabled");
+		return -ENOEXEC;
+	}
+
+	tid = k_thread_create(&event_mon, event_mon_stack,
```

Comment:
Would it work if the user keeps toggling "on" and "off" commands? According to the docs, respawning a thread which terminated is not officially supported (see note):
https://docs.zephyrproject.org/latest/reference/kernel/threads/index.html#thread-termination

Reply:
Seems to work ok, tried with sam_e70_xplained and mimxrt1050_evk boards. Toggling logging did not cause any issues. I would not want to create and run the thread if we do not need it. I also did not want to call the shell APIs from net_mgmt event callback as it has a limited stack.

Reply:
The thread documentation does not really say that respawn is not supported, the notice window just says `The kernel does not currently make any claims regarding an application’s ability to respawn a thread that terminates.` so I read it that it might or might not work ok (just wondering when it could fail).

Reply:
If it works then I guess it's ok, perhaps I've exaggerated the note in the docs. Still IMO it's a bit confusing to understand what exactly `The kernel does not currently make any claims` means and when would this work and when not.

Reply:
> what exactly `The kernel does not currently make any claims` means

@andrewboie do you have any information about the thread termination comment in the documentation?

Reply:
hmm I am not sure. I'm trying to find where this came from in the git history (difficult due to all the reorganization PRs against doc/) and I think this might be very old.

We do respawn threads all the time, re-using their thread objects, in ztest. There shouldn't be any cases where respawning a thread over and over would eventually lead to some kind of resource exhaustion.

I also know that there exist some issues in SMP with respect to thread aborts racing with other uses of thread objects, particularly with respect to threads being woken up from wait queues. There's a bug open on this that I am working on. I don't think there are any races like this on uniprocessor, AFAIK uniprocessor is rock-solid.

One thing in the kernel that sticks out about threads is that even though threads have a lifecycle, we don't have a thread id abstraction like PIDs in Linux. Or we do with `k_tid_t`, but it's just the thread pointer. Things can get sketchy if a `struct k_thread` object gets released to a heap or otherwise re-purposed to some other kind of memory after the thread exits, and someone else tries to use the thread object pointer. You can only really guarantee that a thread is running by holding `sched_spinlock`, otherwise it could potentially abort at any time. I have recently been thinking there may need to be an abstraction here to make this safer, but there are of course footprint concerns. This is all hand-wavy and needs more thought and I just mention it in passing, things have been this way since Viper days AFAIK.

For the purposes of this PR, what you are doing looks fine. `event_mon` isn't re-used as some kind of other memory so as long as `event_mon` isn't already running when you call `k_thread_create()` on it you will be OK. I recently put some assertions in place to catch this.

copying @andyross in case he has any comments or I mis-stated anything.

Reply:
Having said all that, filing a bug against that bit of documentation for further investigation seems worthwhile.

Reply:
Yeah, that doc isn't really correct.  Spawning a thread is just repurposing some memory as a k_thread struct and it will fully initialize it.  Once terminated, that struct is just memory again and can be respawned.  There are race conditions there with double-spawns and aborts, etc...  But the statement in the docs doesn't correctly capture the issue.